File: ballast-core/npm/src/domain/events/vessel-state-changed.ts
import { EventBase } from "../../messaging/event-base";
import { Game } from "../models/game";
import { Vessel } from "../models/vessel";

export class VesselStateChangedDomainEvent extends EventBase {

    public static readonly id: string = 'VesselStateChangedDomainEvent';
    public get id() {
        return VesselStateChangedDomainEvent.id;
    }

    public readonly vessel: Vessel;
    public readonly gameId: string; 

    private constructor(eventDateIsoString: string, gameId: string, vessel: Vessel) {
        super(eventDateIsoString);
        this.gameId = gameId;
        this.vessel = vessel;
    }

    public static fromVesselInGame(game: Game, vessel: Vessel) {
        return new VesselStateChangedDomainEvent(
            EventBase.getDateIsoString(),
            game.id,
            vessel
        );
    }

}

File: ballast-core/dotnet/src/application/events/VesselStateChanged.cs
using System;
using Ballast.Core.Application.Models;
using Ballast.Core.Messaging;

namespace Ballast.Core.Application.Events
{
    public class VesselStateChangedEvent : EventBase, IApplicationEvent  
    {

        public static string GetId() => nameof(VesselStateChangedEvent);
        public override string Id => GetId();

        public Guid GameId { get; private set; }
        public VesselDto Vessel { get; private set; }

        private VesselStateChangedEvent() { } // For model binding / deserialization
        private VesselStateChangedEvent(string eventDateIsoString, Guid gameId, VesselDto vessel) : base(eventDateIsoString) 
        {
            GameId = gameId;
            Vessel = vessel;
        }

        public static VesselStateChangedEvent FromVesselInGame(Guid gameId, VesselDto vessel) =>
            new VesselStateChangedEvent(
                EventBase.GetDateIsoString(),
                gameId, 
                vessel
            );

    }
}